### PR TITLE
Fix unfold2d and add test

### LINKED ISSUE #442: `setStorage` Runtime error `from opacus.utils.tensor_utils.unfold2d`
## 🐛 Bug

The function `unfold2d` throws `setStorage` runtime error when using a `Conv2d` with certain kernel and padding sizes.

## To Reproduce

1. Create activation tensor `[16, 3, 32, 32]`
2. Call opacus.utils.tensor_utils.unfold2d with `activation` and
    - kernel_size = (3,1)
    - padding = (1, 0)
    - stride = (1, 1)
    - dilation = (1,1)

Stack trace
```
[/content/opacus/opacus/utils/tensor_utils.py](https://localhost:8080/#) in unfold2d(input, kernel_size, padding, stride, dilation)
    141     ]
    142     out = input.as_strided(
--> 143         shape + [kernel_size[0], kernel_size[1], H_effective, W_effective], strides
    144     )
    145 

RuntimeError: setStorage: sizes [16, 3, 3, 1, 32, 32], strides [3264, 1088, 34, 1, 34, 1], storage offset 0, and itemsize 4 requiring a storage size of 209160 are out of bounds for storage of size 208896
```

RuntimeError: setStorage: sizes [16, 3, 3, 1, 32, 32], strides [3264, 1088, 34, 1, 34, 1], storage offset 0, and itemsize 4 requiring a storage size of 209160 are out of bounds for storage of size 208896

## Expected behaviour

Not throwing an error?

## Environment
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A
OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A
Python version: 3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19041-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3050 Ti Laptop GPU
Nvidia driver version: 516.40
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h59b6b97_2  
[conda] mkl                       2021.4.0           haa95532_640  
[conda] mkl-service               2.4.0            py38h2bbff1b_0  
[conda] mkl_fft                   1.3.1            py38h277e83a_0  
[conda] mkl_random                1.2.2            py38hf11a4ad_0  
[conda] numpy                     1.22.3           py38h7a0a035_0  
[conda] numpy-base                1.22.3           py38hca35cd5_0  
[conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.11.0               py38_cu113    pytorch
[conda] torchvision               0.12.0               py38_cu113    pytorch

## Additional context

This error shows up when DP training a Conv2D with similar kernel size and padding.  
Oddly there's no error when using the reverse sizes.

Meaning   `kernel_size = (1,3)` and  `padding = (0, 1)` does not throw any errors.

So I'm guessing there is a mathematical mistake in `unfold2d`.  but I can't make heads or tails of.

### PR DESCRIPTION
**Pull request by alexandresablayrolles**
Fixes #442.
TL;DR: F.pad has an unintuitive syntax (we need to indicated padding from last dimensions to first dimensions).
This PR fixes that in unfold2d and adds tests with non-symmetric pad/dilation/kernel_size/stride.

### CODE REVIEW

File: opacus/utils/tensor_utils.py [hunk 1]
```diff
@@ -130,7 +130,8 @@ def unfold2d(
     W_effective = (
         W + 2 * padding[1] - (kernel_size[1] + (kernel_size[1] - 1) * (dilation[1] - 1))
     ) // stride[1] + 1
-    input = F.pad(input, (padding[0], padding[0], padding[1], padding[1]))
+    # F.pad's first argument is the padding of the *last* dimension
```
Comment (karthikprasad) [inline opacus/utils/tensor_utils.py:133]: 🙃

**Review by karthikprasad — APPROVED** [review]

**Comment by karthikprasad** [comment]
Please fix lint before merging.

**Comment by ffuuugor** [comment]
Looks great, thanks for the fix!

Can we also add more inputs to `test_conv2d` (original grad sample test) so it would have caught the issue - I think it's a gap that we don't check the layer with 2d inputs for kernel/padding/stride/dilation

**Review by karthikprasad — APPROVED** [review]
Thanks. LGTM. Please `isort` before merging.
